mlir/lib/Dialect/Rock/Transforms/SugarToLoops.cpp

```diff
@@ -888,6 +888,42 @@ static Value computeMemRefNumElements(OpBuilder &b, Location loc,
   return result;
 }
 
+/// Atomic add for a scalar fp16. Using the CAS loop (atomicRMWOp) alternative
+/// is significantly slower so we extend the scalar in a vector and use the
+/// buffer_atomic_add_fp16 instead. We have to take care of the alignment
+/// manually
+static void atomicFp16AddAligned(OpBuilder &b, Location loc, Value data,
+                                 Value dest, ArrayRef<Value> coords) {
+  const bool useBufferOobChecks = true;
+  // Useful consts
+  Value zero = b.create<arith::ConstantIntOp>(loc, 0, 32);
+  Value one = b.create<arith::ConstantIntOp>(loc, 1, 32);
+  Value two = b.create<arith::ConstantIntOp>(loc, 2, 32);
+
+  // Extended packed data to use with the intrinsic
+  Value dataExt =
+      createZeroConstantOp(b, loc, vectorTypeOrSelf(b.getF16Type(), 2));
+  Value dataExt0 = b.create<vector::InsertElementOp>(loc, data, dataExt, zero);
+  Value dataExt1 = b.create<vector::InsertElementOp>(loc, data, dataExt, one);
+
+  // Manual alignment logic : if (addr % 2 != 0) step{AddressData}Back
+  Value stepBack = b.create<arith::SubIOp>(loc, coords[2], one);
+  Value alignment = b.create<arith::RemUIOp>(loc, coords[2], two);
+  Value notAligns =
```

Comment:
As a note: when `threadwise_write_all` gets lowered, we're doing a vectorization query. That vectorization query also implicitly gives the alignment (I can probably adjust my new vectorization API to explicitly pass the final computed alignment back). So we only need this logic when we know we'll be in odd coordinates.

Reply:
... that might be implicit in the fact that we've gone down this branch, never mind

Reply:
I think you'll need to find the last non-unit dimension

Consider, for example, `store %buf[%x, %c0] : memref<4x1xf16>`

Reply:
Ping on that

Reply:
Done. I also added a couple of tests.